compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -0,0 +1,336 @@
+      transformApply(tree, transform(tree.fun))
+
+    private inline def transformApply(tree: Apply, transformedFun: Tree)(using Context): Apply =
+      cpy.Apply(tree)(transformedFun, transform(tree.args))
+
+    private inline def instrumentCases(cases: List[CaseDef])(using Context): List[CaseDef] =
+      cases.map(instrumentCaseDef)
+
+    private def instrumentCaseDef(tree: CaseDef)(using Context): CaseDef =
+      val pat = tree.pat
+      val guard = tree.guard
+      val friendlyEnd = if guard.span.exists then guard.span.end else pat.span.end
+      val pos = tree.sourcePos.withSpan(tree.span.withEnd(friendlyEnd)) // user-friendly span
+      // ensure that the body is always instrumented by inserting a call to Invoker.invoked at its beginning
+      val instrumentedBody = instrument(transform(tree.body), pos, false)
+      cpy.CaseDef(tree)(tree.pat, transform(tree.guard), instrumentedBody)
+
+    /** Records information about a new coverable statement. Generates a unique id for it.
+      * @return the statement's id
+      */
+    private def recordStatement(tree: Tree, pos: SourcePosition, branch: Boolean)(using ctx: Context): Int =
+      val id = statementId
+      statementId += 1
+      val statement = new Statement(
+        source = ctx.source.file.name,
+        location = Location(tree),
+        id = id,
+        start = pos.start,
+        end = pos.end,
+        line = pos.line,
+        desc = tree.source.content.slice(pos.start, pos.end).mkString,
+        symbolName = tree.symbol.name.toSimpleName.toString,
+        treeName = tree.getClass.getSimpleName.nn,
+        branch
+      )
+      coverage.addStatement(statement)
+      id
+
+    private inline def syntheticSpan(pos: SourcePosition): Span = pos.span.toSynthetic
+
+    /** Shortcut for instrument(tree, tree.sourcePos, branch) */
+    private inline def instrument(tree: Tree, branch: Boolean = false)(using Context): Tree =
+      instrument(tree, tree.sourcePos, branch)
+
+    /** Instruments a statement, if it has a position. */
+    private def instrument(tree: Tree, pos: SourcePosition, branch: Boolean)(using Context): Tree =
+      if pos.exists && !pos.span.isZeroExtent then
+        val statementId = recordStatement(tree, pos, branch)
+        insertInvokeCall(tree, pos, statementId)
+      else
+        tree
+
+    /** Instruments the body of a DefDef. Handles corner cases. */
+    private def instrumentBody(parent: DefDef, body: Tree)(using Context): Tree =
+      /* recurse on closures, so that we insert the call at the leaf:
+
+         def g: (a: Ta) ?=> (b: Tb) = {
+           // nothing here             <-- not here!
+           def $anonfun(using a: Ta) =
+             Invoked.invoked(id, DIR)  <-- here
+             <userCode>
+           closure($anonfun)
+         }
+      */
+      body match
+        case b @ Block((meth: DefDef) :: Nil, closure: Closure) if meth.symbol == closure.meth.symbol =>
```

Comment:
Since only contextual closures need to be special-cased (cf `Config.flattenContextFunctionResults`), we can special-case them:
```suggestion
        case b @ Block((meth: DefDef) :: Nil, closure: Closure)
        if meth.symbol == closure.meth.symbol && defn.isContextFunctionType(body.tpe) =>
```